lib/Bitcode/Reader/BitstreamReader.cpp

```diff
@@ -156,9 +182,178 @@ void BitstreamCursor::skipRecord(unsigned AbbrevID) {
+        report_fatal_error("Array element type can't be an Array or a Blob");
+
+      // Read all the elements a little faster.
+      {
+        BitCodeAbbrevOp::Encoding enc = EltEnc.getEncoding();
+        uint64_t encData = 0;
+        if (EltEnc.hasEncodingData())
+          encData = EltEnc.getEncodingData();
+        const unsigned elemSize = (unsigned)encData;
+        if (enc == BitCodeAbbrevOp::Fixed) {
+          assert((unsigned)encData == 8 && (unsigned)encData <= MaxChunkSize);
+          // Special optimization for fixed elements that are 8 bits
+          Vals.resize(NumElts);
+          uint8_t *ptr = Vals.data();
+          unsigned i = 0;
+          constexpr unsigned BytesInWord = sizeof(size_t);
+          // First, read word by word instead of byte by byte
+          for (; NumElts >= BytesInWord; NumElts -= BytesInWord) {
+            const size_t e = Read(BytesInWord * 8);
+            memcpy(ptr+i, &e, sizeof(e));
+            i += BytesInWord;
+          }
+          for (; NumElts; --NumElts)
+            Vals[i++] = (uint8_t)Read(8);
+        }
+        // VBR Could actually be more than 8-bit. It's up to the caller
+        // to make sure no such case will be possible.
+        else if (enc == BitCodeAbbrevOp::VBR) {
+          assert((unsigned)encData <= MaxChunkSize);
+          assert(EltEnc.getLiteralValue() <= 8);
+          for (; NumElts; --NumElts) {
+            Vals.push_back((uint8_t)ReadVBR64(elemSize));
+          }
+        }
+        else if (enc == BitCodeAbbrevOp::Char6) {
+          Vals.reserve(NumElts);
+          assert((unsigned)encData <= MaxChunkSize);
+          for (; NumElts; --NumElts) {
+            Vals.push_back(BitCodeAbbrevOp::DecodeChar6(Read(6)));
+          }
+        }
+        else {
+          llvm_unreachable("Unknown kind of thing");
+        }
+      }
+      continue;
+    }
+
+    assert(Op.getEncoding() == BitCodeAbbrevOp::Blob);
+    // Blob case.  Read the number of bytes as a vbr6.
+    unsigned NumElts = ReadVBR(6);
+    SkipToFourByteBoundary();  // 32-bit alignment
+
+    // Figure out where the end of this blob will be including tail padding.
+    size_t CurBitPos = GetCurrentBitNo();
+    size_t NewEnd = CurBitPos+((NumElts+3)&~3)*8;
+
+    // If this would read off the end of the bitcode file, just set the
+    // record to empty and return.
+    if (!canSkipToPos(NewEnd/8)) {
+      Vals.append(NumElts, 0);
+      NextChar = BitStream->getBitcodeBytes().getExtent();
+      break;
+    }
+
+    // Otherwise, inform the streamer that we need these bytes in memory.
+    const char *Ptr = (const char*)
+      BitStream->getBitcodeBytes().getPointer(CurBitPos/8, NumElts);
+
+    // If we can return a reference to the data, do so to avoid copying it.
+    if (Blob) {
+      *Blob = StringRef(Ptr, NumElts);
+    } else {
+      // Otherwise, unpack into Vals with zero extension.
+      for (; NumElts; --NumElts)
+        Vals.push_back((unsigned char)*Ptr++);
```

Comment:
Could this be `Vals.insert(Vals.end(), (const uint8_t*)Ptr, (const uint8_t*)Ptr + NumElts)`?

Reply:
Yes, but Blob doesn't make up much right now.